#292 FbStatus without second argument

komaz opened the pull request:
Fix for issue #292. As discussed in issue, I'm taking a response code with explanation message from `RsWithStatus` head.

davvd commented:
@komaz I will find a reviewer for your pull requests shortly, thanks for contribution!

davvd commented:
@ggajos it's yours,please review

ggajos commented on src/main/java/org/takes/facets/fallback/FbStatus.java:
```diff
@@ -45,7 +47,46 @@
+                throws IOException {
+                final Response res = new RsWithStatus(req.code());
+                final String message = res
+                    .head().iterator().next().substring("HTTP 1.1 ".length());
```
@komaz that's not looking good, what if HTTP will look like `HTTP/1.1 200 OK`?

komaz replied:
@ggajos in that case `200 OK` will be extracted, and the full message in the body will be something like `200 OK: <message from throwable>`. I assume that's exactly what we need, right? If someone creates fallback for status 200, then probably that's what he or she wants. Or do you mean there's a typo in a string, i.e. it should be `HTTP/1.1` instead of `HTTP 1.1`?

ggajos commented on src/test/java/org/takes/facets/fallback/FbStatusTest.java:
```diff
@@ -102,4 +104,31 @@ public void ignoresDifferentStatus() throws Exception {
     }
+
+    /**
+     * Default response body should contain status code,
```
@komaz wrong comment, http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names

ggajos commented on src/main/java/org/takes/facets/fallback/FbStatus.java:
```diff
@@ -45,7 +47,46 @@
+        this(new Condition<Integer>() {
+            @Override
+            public boolean fits(final Integer status) {
+                return code == status;
```
@komaz autoboxing will be used here, but maybe let's make it visible? .`intValue()` or `equals()`

ggajos commented:
@komaz few comments

komaz commented:
@ggajos thanks for the review! I fixed test case javadoc, added `.intValue()` in condition and fixed a typo in `HTTP/1.1` prefix. Could you please comment what is wrong with extracting only a status and status message from a head?

ggajos commented:
@komaz I think that relying on version number is bad. Maybe we can use some regexp or `space` character so this implementation will still work when we deal with http/1.0. What do you think?

komaz commented:
@ggajos The code relies on a string length (`line.substring("HTTP/1.1 ".length())`), so I guess as long as this project won't respond with something like `HTTP/2.10`, or `HTTP/1.11`, it is safe. I'm afraid using regex like `HTTP/1\\.1\\s+(\\d+\\s?.*)` would make things even less readable and require dealing with matcher groups.

Maybe splitting this long expression to several lines would help? Something like this:
```java
final String first = res.head().iterator().next();
final int prefix = "HTTP/1.1 ".length();
final String message = first.substring("HTTP/1.1 ".length());
```
But in that case there'll be two local variables which are used only once

ggajos commented:
@komaz Current solution is ok I just wondering how we can avoid eventual surprises in future but maybe let's not over engineer this. Thanks for explanations.

ggajos commented:
@rultor merge please

rultor commented:
> @rultor merge please

@ggajos Thanks for your request. @yegor256 Please confirm this.

yegor256 commented on src/main/java/org/takes/facets/fallback/FbStatus.java:
```diff
@@ -45,7 +47,46 @@
+            public Iterator<Response> route(final RqFallback req)
+                throws IOException {
+                final Response res = new RsWithStatus(req.code());
+                final String message = res
```
`message` looks like a one-time variable, see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#one-time-variables

yegor256 commented on src/main/java/org/takes/facets/fallback/FbStatus.java:
```diff
@@ -45,7 +47,46 @@
+                throws IOException {
+                final Response res = new RsWithStatus(req.code());
+                final String message = res
+                    .head().iterator().next().substring("HTTP/1.1 ".length());
```
this is not safe, what if there will be `HTTP/2.0`? I would rather do `.split("\\s", 2)[1]`

yegor256 commented on src/main/java/org/takes/facets/fallback/FbStatus.java:
```diff
@@ -45,7 +47,46 @@
+                    .head().iterator().next().substring("HTTP/1.1 ".length());
+                return Collections.<Response>singleton(
+                    new RsFluent(res)
+                        .withHeader("Content-Type", "text/plain")
```
it's better to use `RsText`

yegor256 commented:
@komaz a few comments from me above

komaz commented:
@yegor256 
- The code relies on a length of `"HTTP/1.1 "` string. Since `"HTTP/2.0 "` has the same length, everything will work as expected. But I agree your solution is better, I've pushed an updated version.
- Can't use `RsText` until #310 is fixed, changed to `RsWithType` and improved a test to check response status code. Or should I wait for #310 fix?

yegor256 commented:
@rultor try to merge

rultor commented:
> @rultor try to merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/3846-104803845)

rultor commented:
> @rultor try to merge

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/3846-104803845) (took me 12min)

davvd commented:
@elenavolokhova please, review this task for compliance with our [quality rules](http://www.teamed.io/qa.html)

davvd commented:
@rultor deploy now pls

rultor commented:
> @rultor deploy now pls

@davvd OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/3846-104998344)

rultor commented:
> @rultor deploy now pls

@davvd Done! FYI, the full log is [here](http://www.rultor.com/t/3846-104998344) (took me 7min)

elenavolokhova commented:
@ggajos 
>yegor256 added a note 2 days ago
this is not safe, what if there will be HTTP/2.0? I would rather do .split("\\s", 2)[1]

And couple more of his comments have no addressee in the beginning. In the next ticket please ask him to put username to all comments even if they are left on 'Files changed' tab.

I'm asking you as a performer of this ticket who is responsible for it's quality.

Please confirm that you will try to fix such issues in your future tickets.

elenavolokhova commented:
@ggajos I need your confirmation in order to complete QA review

ggajos commented:
@elenavolokhova I can't see this message anywhere here `this is not safe, what if there will be HTTP/2.0? I would rather do .split("\s", 2)[1]`

elenavolokhova commented:
@ggajos It is in comments left on 'Files Changed' tab. They are visible here in expandable sections like: Show outdated diff yegor256 commented on an outdated diff 7 days ago

It is not obvious. I know. But it is very important for PR creator to receive notification from that page too.

ggajos commented:
@elenavolokhova ok, thanks, got it now

ggajos commented:
@yegor256 please remember to address all comments directly

elenavolokhova commented:
@davvd Quality is acceptable here.

davvd commented:
> @davvd Quality is acceptable here.

@elenavolokhova thanks a lot, next time everybody should try to make it better

davvd commented:
@ggajos just added 10 mins to @elenavolokhova (for QA), payment ID is `58205785`... Thank you, I have added **22 mins** to your account in payment/transaction "`AP-27333210SG973874N`", time consumed: 95 hours and 25 mins... extra minutes for review comments (c=7)... added +22 to your rating, now it is equal to [+2231](http://www.netbout.com/b/35094?open=rating)
